contracts/TokenLocking.sol

```diff
@@ -0,0 +1,53 @@
+pragma solidity ^0.4.23;
+
+import "../lib/dappsys/math.sol";
+import "./ERC20Extended.sol";
+import "./TokenLockingStorage.sol";
+
+
+contract TokenLocking is TokenLockingStorage, DSMath {
+
+  modifier notLocked(address _token) {
+    require(!userTokenLockChanges[msg.sender][_token].isLocked);
+    _;
+  }
+
+  function deposit(address _token, uint256 _amount) public notLocked(_token) {
+    userDepositedTokens[msg.sender][_token] = add(userDepositedTokens[msg.sender][_token], _amount);
+
+    require(ERC20Extended(_token).transferFrom(msg.sender, address(this), _amount));
+  }
+
+  function withdraw(address _token, uint256 _amount) public notLocked(_token) {
+    userDepositedTokens[msg.sender][_token] = sub(userDepositedTokens[msg.sender][_token], _amount);
+
+    require(ERC20Extended(_token).transfer(msg.sender, _amount));
+  }
+
+  function getUserTokens(address _user, address _token) public view returns (uint256) {
+    return userDepositedTokens[_user][_token];
+  }
+
+  function getUserLockInfo(address _user, address _token) public view returns (uint256, bool) {
+    Lock memory lock = userTokenLockChanges[_user][_token];
+    return (lock.timestamp, lock.isLocked);
+  }
+
+  function lock(address _token) public {
```

Comment:
Why is `lock` separate to `deposit`?

Reply:
Delegating the responsibility to the user for locking their own tokens depends on them knowing when to lock exactly, i.e. they all have to lock before the `startNextRewardPayout` is called and also relies on *all* of them locking their tokens otherwise the `RewardPayoutCycle.totalTokens` (which is calculated as all tokens owned by anyone but the colony itself) will be wrong.

Reply:
I think that users would like to have control of their tokens. We could help the on the front-end so they don't mess up.

You think that `RewardPayoutCycle.totalTokens` should not be `sub(token.totalSupply(), token.balanceOf(address(this)))`, but it should be sum of all deposited tokens?

Reply:
Also, yeah, i could merge `lock` and `deposit`

Reply:
> I think that users would like to have control of their tokens.

The way that I have envisioned this happening, this should always be true. Ultimately, tokens will be able to be locked for two reasons: 

* A payout is live, and they've not claimed or waived their payout yet
* A token-weighted vote they have taken part in has reached the 'reveal' stage.

In both cases, it is completely in the user's control to unlock their tokens. In the first case, they claim their payout or waive their claim to the payout. In the second case, they reveal any votes they committed.

Reply:
In this implementation, users tokens will not be unlocked just by calling `claimNextRewardPayout`, they need to call `unlock` (in TokenLocking contract) themselves.

There are two scenarios we are discussing here:
1) Using counter, where we lock the users tokens and we unlock them, e.g:
- User deposits his tokens
- We call `startNextRewardPayout` (we will lock everyones tokens inside this function)
- User calls `claimNextRewardPayout` (we will unlock users tokens in this function)

2) Using this implementation:
- User deposits token
- **User** locks his tokens by calling `lock` in TokenLocking contract
- We call `startNextRewardPayout` (we dont touch users tokens here)
- User calls `claimNextRewardPayout` (we check if he locked his tokens before reward payout start)
- **User** unlocks his tokens

Reply:
So what are the benefits of doing it this way? To me, it looks like it requires extra transactions by the user. I agree that in general this shouldn't prevent us from doing something - as you have mentioned, the UI can guide the user. But if we can avoid them, it seems to me like we should?

Reply:
Thats a valid point.
There is maybe one exception where this one would be cheaper vs the version with counters:
- when users tokens are locked multiple times, either by reward payouts, voting etc. he would need to unlock them multiple times (increase the counter by claiming, waiving, whatever), wheres here he would finish all the stuff he needs to finish, and at the end unlock the tokens once. Im not saying if this valid enough argument to choose this one over the other, but maybe worth to consider.